fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingThrottle.java

```diff
@@ -0,0 +1,264 @@
+                                    LogMessageKeys.LIMIT, limit,
+                                    LogMessageKeys.MAX_LIMIT, maxLimit).toString());
+                }
+                limit = maxLimit;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    void decreaseLimit(@Nonnull FDBException fdbException,
+                       @Nullable List<Object> additionalLogMessageKeyValues) {
+        limit = Math.max(1, (3 * limit) / 4);
+        if (LOGGER.isInfoEnabled()) {
+            final KeyValueLogMessage message = KeyValueLogMessage.build("Lessening limit of online index build",
+                    LogMessageKeys.ERROR, fdbException.getMessage(),
+                    LogMessageKeys.ERROR_CODE, fdbException.getCode(),
+                    LogMessageKeys.LIMIT, limit,
+                    LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                    LogMessageKeys.INDEXER_ID, common.getUuid()
+                    );
+            if (additionalLogMessageKeyValues != null) {
+                message.addKeysAndValues(additionalLogMessageKeyValues);
+            }
+            LOGGER.info(message.toString(), fdbException);
+        }
+    }
+
+    private void tryToIncreaseLimit(@Nullable Throwable exception) {
+        final int increaseLimitAfter = common.config.getIncreaseLimitAfter();
+        final int maxLimit = common.config.getMaxLimit();
+        if (increaseLimitAfter > 0) {
+            if (exception == null) {
+                successCount++;
+                if (successCount >= increaseLimitAfter && limit < maxLimit) {
+                    increaseLimit();
+                }
+            } else {
+                successCount = 0;
+            }
+        }
+    }
+
+    private void increaseLimit() {
+        final int maxLimit = common.config.getMaxLimit();
+        limit = Math.min(maxLimit, Math.max(limit + 1, (4 * limit) / 3));
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info(KeyValueLogMessage.of("Re-increasing limit of online index build",
+                    LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
+                    LogMessageKeys.INDEXER_ID, common.getUuid(),
+                    LogMessageKeys.LIMIT, limit));
+        }
+    }
+
+    // Finds the FDBException that ultimately caused some throwable or
+    // null if there is none. This can be then used to determine, for
+    // example, the error code associated with this FDBException.
+    @Nullable
+    private FDBException getFDBException(@Nullable Throwable e) {
+        Throwable curr = e;
+        while (curr != null) {
+            if (curr instanceof FDBException) {
+                return (FDBException)curr;
+            } else {
+                curr = curr.getCause();
+            }
+        }
+        return null;
+    }
+
+    @Nonnull
+    <R> CompletableFuture<R> runAsync(@Nonnull final Function<FDBRecordStore, CompletableFuture<R>> function,
```

Comment:
I'm assuming this was just copied from somewhere else, but this is a pretty complicated method... no concrete suggestions, and not necessarily arguing it should be cleaned up now.